### PR TITLE
Missing bytecode of sources for dependece scan is no longer fatal

### PR DESCRIPTION
**Pull request by judovana**
If class is compiled, whcih was originally not on original CP, compilation fails,  because no bytecode is returned from ClassesProvider and thus objectweb asm is trying to read in byte[0] or in null.

The PR shoudl be making this case non fatal. See the example of our JRD friednd:
```
package [org.jrd.frontend.frame.main.renderer](http://org.jrd.frontend.frame.main.renderer/);
public class VmListRenderer {
    
}
```
will compile  (against JRD)

however
```
public class a {
}

```
not

The exception was
```
removing -target and 11
removing -source and 11
compiler args: -source 11 -target 11 -verbose
 restoring agent verified on : localhost:10900
 restored agent verified on : localhost:10900
Processing request. VM ID: 15866, PID: 15866, action: BYTES, port: 10900
java.lang.RuntimeException: Class org.jrd.frontend.frame.main.renderer.a not found in loaded classes.
    at org.jrd.agent.InstrumentationProvider.findClass(InstrumentationProvider.java:90)
    at org.jrd.agent.InstrumentationProvider.findClassBody(InstrumentationProvider.java:79)
    at org.jrd.agent.AgentActionWorker.sendByteCode(AgentActionWorker.java:230)
    at org.jrd.agent.AgentActionWorker.writeToStreamBasedOnLine(AgentActionWorker.java:131)
    at org.jrd.agent.AgentActionWorker.executeRequest(AgentActionWorker.java:95)
    at org.jrd.agent.AgentActionWorker.<init>(AgentActionWorker.java:44)
    at org.jrd.agent.ConnectionDelegator.run(ConnectionDelegator.java:77)
java.lang.RuntimeException: Agent returned error in response header: java.lang.RuntimeException: Class org.jrd.frontend.frame.main.renderer.a not found in loaded classes.
    at org.jrd.backend.communication.Communicate.readResponse(Communicate.java:120)
    at org.jrd.backend.communication.CallDecompilerAgent.submitRequest(CallDecompilerAgent.java:48)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:170)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:149)
    at org.jrd.backend.core.DecompilerRequestReceiver.getByteCodeAction(DecompilerRequestReceiver.java:240)
    at org.jrd.backend.core.DecompilerRequestReceiver.processRequest(DecompilerRequestReceiver.java:84)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.submitRequest(DecompilationController.java:746)
    at org.jrd.backend.data.cli.Lib.obtainClass(Lib.java:286)
    at org.jrd.backend.communication.RuntimeCompilerConnector$JrdClassesProvider.getClass(RuntimeCompilerConnector.java:56)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getSingleClassFromRunningVmCatched(DecompilationController.java:544)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getClassItself(DecompilationController.java:533)
    at org.jrd.frontend.frame.main.decompilerview.dummycompiler.ClassesAndMethodsProviderBasedClassesProvider.getClass(ClassesAndMethodsProviderBasedClassesProvider.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:86)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
java.lang.RuntimeException: Agent returned error response 'java.lang.RuntimeException: Class org.jrd.frontend.frame.main.renderer.a not found in loaded classes.' for request 'BYTES\norg.jrd.frontend.frame.main.renderer.a'.
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:175)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:149)
    at org.jrd.backend.core.DecompilerRequestReceiver.getByteCodeAction(DecompilerRequestReceiver.java:240)
    at org.jrd.backend.core.DecompilerRequestReceiver.processRequest(DecompilerRequestReceiver.java:84)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.submitRequest(DecompilationController.java:746)
    at org.jrd.backend.data.cli.Lib.obtainClass(Lib.java:286)
    at org.jrd.backend.communication.RuntimeCompilerConnector$JrdClassesProvider.getClass(RuntimeCompilerConnector.java:56)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getSingleClassFromRunningVmCatched(DecompilationController.java:544)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getClassItself(DecompilationController.java:533)
    at org.jrd.frontend.frame.main.decompilerview.dummycompiler.ClassesAndMethodsProviderBasedClassesProvider.getClass(ClassesAndMethodsProviderBasedClassesProvider.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:86)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
java.lang.RuntimeException: Classes couldn't be loaded.Do you have agent configured?On JDK 9 and higher, did you run the target process with '-Djdk.attach.allowAttachSelf=true'? Or maybe agent is not loaded? Or bad agent?
    at org.jrd.backend.data.cli.Lib.obtainClass(Lib.java:290)
    at org.jrd.backend.communication.RuntimeCompilerConnector$JrdClassesProvider.getClass(RuntimeCompilerConnector.java:56)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getSingleClassFromRunningVmCatched(DecompilationController.java:544)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getClassItself(DecompilationController.java:533)
    at org.jrd.frontend.frame.main.decompilerview.dummycompiler.ClassesAndMethodsProviderBasedClassesProvider.getClass(ClassesAndMethodsProviderBasedClassesProvider.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:86)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
Attempting to init the class and load again
 restoring agent verified on : localhost:10900
 restored agent verified on : localhost:10900
Processing request. VM ID: 15866, PID: 15866, action: INIT_CLASS, port: 10900
java.lang.ClassNotFoundException: org.jrd.frontend.frame.main.renderer.a
    at java.base/…581)
    at java.base/…178)
    at java.base/…522)
    at java.base/…ive Method)
    at java.base/…315)
    at org.jrd.agent.AgentActionWorker$4.run(AgentActionWorker.java:284)
    at org.jrd.agent.AgentActionWorker.executeParametrisedNoReturnCommand(AgentActionWorker.java:270)
    at org.jrd.agent.AgentActionWorker.initClass(AgentActionWorker.java:281)
    at org.jrd.agent.AgentActionWorker.writeToStreamBasedOnLine(AgentActionWorker.java:146)
    at org.jrd.agent.AgentActionWorker.executeRequest(AgentActionWorker.java:95)
    at org.jrd.agent.AgentActionWorker.<init>(AgentActionWorker.java:44)
    at org.jrd.agent.ConnectionDelegator.run(ConnectionDelegator.java:77)
java.lang.RuntimeException: Agent returned error in response header: java.lang.ClassNotFoundException: org.jrd.frontend.frame.main.renderer.a
    at org.jrd.backend.communication.Communicate.readResponse(Communicate.java:120)
    at org.jrd.backend.communication.CallDecompilerAgent.submitRequest(CallDecompilerAgent.java:48)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:170)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:149)
    at org.jrd.backend.core.DecompilerRequestReceiver.getNoReplyValue(DecompilerRequestReceiver.java:225)
    at org.jrd.backend.core.DecompilerRequestReceiver.getInitAction(DecompilerRequestReceiver.java:220)
    at org.jrd.backend.core.DecompilerRequestReceiver.processRequest(DecompilerRequestReceiver.java:77)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.submitRequest(DecompilationController.java:746)
    at org.jrd.backend.data.cli.Lib.initClassNoThrow(Lib.java:61)
    at org.jrd.backend.data.cli.Lib.initClass(Lib.java:49)
    at org.jrd.backend.communication.RuntimeCompilerConnector$JrdClassesProvider.getClass(RuntimeCompilerConnector.java:61)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getSingleClassFromRunningVmCatched(DecompilationController.java:544)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getClassItself(DecompilationController.java:533)
    at org.jrd.frontend.frame.main.decompilerview.dummycompiler.ClassesAndMethodsProviderBasedClassesProvider.getClass(ClassesAndMethodsProviderBasedClassesProvider.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:86)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
java.lang.RuntimeException: Agent returned error response 'java.lang.ClassNotFoundException: org.jrd.frontend.frame.main.renderer.a' for request 'INIT_CLASS\norg.jrd.frontend.frame.main.renderer.a'.
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:175)
    at org.jrd.backend.core.DecompilerRequestReceiver.getResponse(DecompilerRequestReceiver.java:149)
    at org.jrd.backend.core.DecompilerRequestReceiver.getNoReplyValue(DecompilerRequestReceiver.java:225)
    at org.jrd.backend.core.DecompilerRequestReceiver.getInitAction(DecompilerRequestReceiver.java:220)
    at org.jrd.backend.core.DecompilerRequestReceiver.processRequest(DecompilerRequestReceiver.java:77)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.submitRequest(DecompilationController.java:746)
    at org.jrd.backend.data.cli.Lib.initClassNoThrow(Lib.java:61)
    at org.jrd.backend.data.cli.Lib.initClass(Lib.java:49)
    at org.jrd.backend.communication.RuntimeCompilerConnector$JrdClassesProvider.getClass(RuntimeCompilerConnector.java:61)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getSingleClassFromRunningVmCatched(DecompilationController.java:544)
    at org.jrd.frontend.frame.main.decompilerview.DecompilationController.getClassItself(DecompilationController.java:533)
    at org.jrd.frontend.frame.main.decompilerview.dummycompiler.ClassesAndMethodsProviderBasedClassesProvider.getClass(ClassesAndMethodsProviderBasedClassesProvider.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:86)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
Init of class 'org.jrd.frontend.frame.main.renderer.a' failed, not obtaining.
java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 0
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:142)
    at org.jrd.frontend.frame.overwrite.OverwriteClassDialog$CompilationWithResult.run(OverwriteClassDialog.java:562)
    at org.jrd.frontend.frame.main.decompilerview.QuickCompiler$2.run(QuickCompiler.java:55)
    at java.base/…829)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 0
    at org.objectweb.asm.ClassReader.readShort(ClassReader.java:3608)
    at org.objectweb.asm.ClassReader.<init>(ClassReader.java:197)
    at org.objectweb.asm.ClassReader.<init>(ClassReader.java:180)
    at org.objectweb.asm.ClassReader.<init>(ClassReader.java:166)
    at io.github.mkoncek.classpathless.util.BytecodeExtractor.extractOuterClass(BytecodeExtractor.java:216)
    at io.github.mkoncek.classpathless.util.BytecodeExtractor.extractFullClassGroup(BytecodeExtractor.java:268)
    at io.github.mkoncek.classpathless.util.BytecodeExtractor.extractDependenciesImpl(BytecodeExtractor.java:311)
    at io.github.mkoncek.classpathless.util.BytecodeExtractorAccessor.extractDependenciesImpl(BytecodeExtractorAccessor.java:28)
    at io.github.mkoncek.classpathless.impl.CompilerJavac.compileClass(CompilerJavac.java:87)
    ... 3 more
Attempt to compile failed with - java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 0.
```

### CODE REVIEW

**Comment by mkoncek** [comment]
I reviewed the PR.
1) It is mostly adding null checks and zero-length checks which I dislike. I am questioning even catching exceptions from `BytecodeExtractor`. I would rather have a strong specified contract in the API which says that the `ClassesProvider` never returns `null` and that the collection never contains a `null` object. I **think** this may be solved completely by modifying JRD without touching CPLC code.
\
Nevertheless can you try first inspecting all the return paths of JRD's impl of `getClass` and make sure it always returns valid objects? (Reminder that `Collections.emptyList()` exists.)
2) Please check for typos when writing commit messages.
3) The logged message should start with a capital letter and form a more-less plain English sentence.

**Comment by judovana** [comment]
> @judovana I reviewed the PR.
> 
>     1. It is mostly adding null checks and zero-length checks which I dislike. I am questioning even catching 

Its always better to do a null check,then catch NPE. Its so much better that een findbugs and simialr tools will rebuuke you for.
I dont like the zero check too. BUt htat shoudl have been solved much more early in development. IIRC, it was not easy to provide. 
 - on one side yo want to provide as much as possible
 - on second you need to let user know that some part fo that failed.
 => that is literally making the possibility of empty return. 

> exceptions from `BytecodeExtractor`. I would rather have a strong specified contract in the API which says that the `ClassesProvider` never returns `null` and that the collection never contains a `null` object. I **think** this may be solved completely by modifying JRD without touching CPLC code.

The contract is not here. For many reasons, and JRD is using those, the JRD's implemetatins of CLassesProvider return empty arrays. Its to late to change that. 

Even if cotnrract is reworked,  then there woudl be exception or null, so somethign for CplC to adapt to.

the check fior null member is most likely not necessary. If you wish, I will remove it. On contrary, in contrat of Collection is, that it can contain null members.

>        
>        Nevertheless can you try first inspecting all the return paths of JRD's impl of `getClass` and make sure it always returns valid objects? (Reminder that `Collections.emptyList()` exists.)

It returns valid objects, but that not necessarily need to mean that asmtools will always read it. so the catch is necessary if nothing else.
> 
>     2. Please check for typos when writing commit messages.
> 
>     3. The logged message should start with a capital letter and form a more-less plain English sentence.

Gosh.  By what contract ?-)

**Comment by mkoncek** [comment]
@judovana At first I wanted to add comments but I decided to edit your PR, see https://github.com/mkoncek/classpathless-compiler/compare/master...noDepsNoFatal

**Comment by judovana** [comment]
"     * @return The bytecode files of all found classes. Must not be null." no metter how many times you will write "must not be null" , at the end it solves nothing.  The implementation is not under your influence, and no meter how hard anybbody tries, exception or null may occurs. It remains disputable whether to let such buggy behavior just logged and not blocking the rest of the library work, or to let it crash.

Considering the nature of this method, that it is actually just a companion, which helps future real work, then I'm still inclining to the  error-proof  approach.

in https://github.com/mkoncek/classpathless-compiler/commit/226a24a61310aec08917e4504213c28f51c2819b - is there something heavil changed? I wanted to rework the modified block  a bit, so it is much mor readable, and the null/empty checks are not so insulting/intrusive

There is no reason for "
```
-@SuppressFBWarnings(value = {"EQ_DOESNT_OVERRIDE_EQUALS"}, justification = "intentionally using equals from superclass")
+@SuppressFBWarnings(value = {"EQ_DOESNT_OVERRIDE_EQUALS"},
 +   justification = "intentionally using equals from superclass")
```
Or is there?

I will base my today refactoring on yours version.

**Comment by mkoncek** [comment]
> in https://github.com/mkoncek/classpathless-compiler/commit/226a24a61310aec08917e4504213c28f51c2819b - is there something heavil changed?

No, the commit is more-less identical, except i use `var` wherever possible and I changed the log Level in some cases and the log messages.

**Comment by judovana** [comment]
> > in [226a24a](https://github.com/mkoncek/classpathless-compiler/commit/226a24a61310aec08917e4504213c28f51c2819b) - is there something heavil changed?
> 
> No, the commit is more-less identical, except i use `var` wherever possible and I changed the log Level in some cases and the log messages.

the var is EEEVIIILLLL!!!!!! So much evil... Ut yah, CplC is unifed on that.

**Comment by judovana** [comment]
btw I was abit affraid that the umodificable list returned from stream will not work for compielr api, but it works fine.

**Comment by mkoncek** [comment]
Which one exactly?

File: impl/src/main/java/io/github/mkoncek/classpathless/impl/CompilerJavac.java [hunk 1]
```diff
@@ -78,18 +79,11 @@ public Collection<IdentifiedBytecode> compileClass(
                     Stream.of(javaSourceFiles).map(jsf -> jsf.getClassIdentifier().getFullName())
                     .collect(Collectors.toUnmodifiableList()));
 
-            Collection<InMemoryJavaSourceFileObject> compilationUnits = new ArrayList<>();
-            var availableClasses = new TreeSet<String>();
+            var compilationUnits = Arrays.stream(javaSourceFiles)
+                    .map( source -> new InMemoryJavaSourceFileObject(source))
```
Comment (judovana) [inline impl/src/main/java/io/github/mkoncek/classpathless/impl/CompilerJavac.java:83]: this one

**Comment by mkoncek** [comment]
Looks like Collectors::toList returns an ArrayList.
Moreover, compilationUnits is not modified further.

**Pull request merged**